build EFI image for up_squared as post build step

nashif opened the pull request:
- zepfi.py: minor cleanup
- build EFI image for up_squared as post build step

nashif committed:
x86: zefi: support arguments and make compatible with windows

Add argument parsing and use os.path.join where possible to support
building on windows.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
boards: up_squared: add a post build target

Build an EFI image as a post build step.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

andrewboie commented:
can we please get a summary description for this PR ?

jhedberg reviewed (approved):
Verified to work fine, thanks! This is much cleaner than having to manually run and the getting the temporary files in whatever directory you run zefi.py from rather than the build directory where they really belong.

jhedberg commented:
Actually this seems to have exposed another incorrect "multiple of 8" assumption:
```
Traceback (most recent call last):
  File "/workdir/zephyr/arch/x86/zefi/zefi.py", line 159, in 
    build_elf(args.elf_file)
  File "/workdir/zephyr/arch/x86/zefi/zefi.py", line 53, in build_elf
    assert (h.p_filesz % 8) == 0
AssertionError
ninja: build stopped: subcommand failed.
```

So a similar fix as I did for p_memsz is also needed for p_filesz

jhedberg commented:
> Actually this seems to have exposed another incorrect "multiple of 8" assumption

@nashif I *think* all that's needed is to remove that assert, since the "round up" that I added operates on the result of `h.p_memsz - h.p_filesz`, i.e. it already doesn't care if p_filesz is a multiple of 8 or not.

jhedberg commented on arch/x86/zefi/zefi.py:
```diff
@@ -1,134 +1,158 @@
+            sd = seg.data()
+            verbose("%d bytes of data at 0x%x, data offset %d"
+                % (len(sd), h.p_vaddr, len(data_blob)))
+            data_segs.append((h.p_vaddr, len(sd) / 8, len(data_blob) / 8))
```
I just realised that `len(sd)` probably matches with `h.p_filesz` since there's the `assert len(seg.data()) == h.p_filesz`, so if that's no longer a multiple of 8 then the above division would leave the last few bytes out.

jhedberg replied:
..meaning the second parameter of `data_segs.append()` probably needs to be rounded up, or the left-over bytes should be copied to a separate 8-byte buffer and a second `append()` call done for them. @andyross thoughts?

nashif replied:
please push a patch that makes those changes to this PR.

andyross replied:
Yeah.  The reason those asserts were there is that the C implementation of the segment copies is written to use qwords, and I was "pretty sure but not certain" that the ELF output from our linker scripts was going to be padded appropriately, so I threw some asserts in there just for safety.   Obviously that's not the case and we need to pad.

jhedberg replied:
@andyross thanks for clarifying. One more question: should `seg.data()` be considered read-only, thereby requiring a separate buffer to copy the overflowing bytes to, or is it ok to directy extend it with the padding?

jhedberg replied:
.. Or maybe I'm just confusing myself since you're checking the length of `sd` but copying from `h.p_vaddr`. I'm just not sure where exactly the simplest way to deal with the padding is.

jhedberg replied:
@andyross never mind my previous babbling :) I see that what you're appending is just bookkeeping information and the actual copying of data happens later. However the question still remains, since I guess we can't just request to copy extra bytes from an address if the data isn't there (unlike the zero_segs). Would it perhaps make more sense to "fix" the linker scripts so zefi doesn't need to do the padding?

andyross replied:
It probably would.  This tool has the architecture it does simply because I wanted to eliminate variables.  I knew the existing ELF file, loaded as it directs, works.  So I wanted to duplicate that.  Modifying the linker scripts simultaneously would be changing too many things at once. :)

andyross replied:
seg.data() is an object owned by the elftools parser.  I wouldn't modify it directly for hygine reasons, but doing so might work.  I have no idea ;)

jhedberg reviewed (changes requested):
Marking as request changes, see my inline comment

nashif commented on arch/x86/zefi/zefi.py:
```diff
@@ -1,134 +1,158 @@
+    #
+    # Open the ELF file up and find our entry point
+    #
+    fp = open(elf_file, "rb")
```
I just reworked the code, this was already like this :)

ceolin replied:
fair enough :)

andyross replied:
FWIW: the original wasn't in a function, so the fact that objects get closed on completion was implicit.   Personally I like that style (for exactly this reason, mostly), but I know it isn't typical in the python world.

Commit:
arch: x86: zefi: Reduce data section alignment requirement from 8 to 4

It's not safe to assume that the data section is 8-byte aligned.
Assuming 4-byte alignment seems to work however, and results in
simpler code than arbitrary alignment support.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
@nashif I just pushed two patches to re-introduce the original asserts and instead try to do the padding through the linker script. Let's see if CI is happy with that.

jhedberg commented:
Seems like`tests/application_development/libcxx` with `CONFIG_EXCEPTIONS=y` still fails :(

I wonder if it's even enough to ensure alignment in the arch-specific linker script, since there seems to be many sections not affected by it which are not appropriately aligned:
```
Section Headers:
  [Nr] Name              Type             Address           Offset
       Size              EntSize          Flags  Link  Info  Align
  [ 0]                   NULL             0000000000000000  00000000
       0000000000000000  0000000000000000           0     0     0
  [ 1] .locore           PROGBITS         0000000000008000  00001000
       0000000000007000  0000000000000000 WAX       0     0     4096
  [ 2] text              PROGBITS         0000000000100000  00008000
       0000000000021000  0000000000000000  AX       0     0     16
  [ 3] initlevel         PROGBITS         0000000000121000  00029000
       0000000000000090  0000000000000000   A       0     0     8
  [ 4] ctors             PROGBITS         0000000000121090  00029090
       0000000000000010  0000000000000000  WA       0     0     1
  [ 5] init_array        INIT_ARRAY       00000000001210a0  000290a0
       0000000000000010  0000000000000008  WA       0     0     8
  [ 6] log_const_section PROGBITS         00000000001210b0  000290b0
       0000000000000010  0000000000000000   A       0     0     16
  [ 7] rodata            PROGBITS         00000000001210c0  000290c0
       0000000000005868  0000000000000000   A       0     0     32
  [ 8] .gcc_except_table PROGBITS         0000000000126928  0002e928
       0000000000000104  0000000000000000   A       0     0     4
  [ 9] bss               NOBITS           0000000000127000  0002ea2c
       0000000000c03e58  0000000000000000  WA       0     0     4096
  [10] noinit            NOBITS           0000000000d2b000  0002ea2c
       0000000000014000  0000000000000000  WA       0     0     4096
  [11] datas             PROGBITS         0000000000d3f000  0002ea40
       0000000000001008  0000000000000000  WA       0     0     32
  [12] devices           PROGBITS         0000000000d40008  0002fa48
       00000000000000a0  0000000000000000  WA       0     0     8
  [13] k_mutex_area      PROGBITS         0000000000d400a8  0002fae8
       0000000000000020  0000000000000000  WA       0     0     8
  [14] k_sem_area        PROGBITS         0000000000d400c8  0002fb08
       0000000000000018  0000000000000000  WA       0     0     8
  [15] .comment          PROGBITS         0000000000000000  0002fb20
       0000000000000032  0000000000000001  MS       0     0     1
  [16] .debug_aranges    PROGBITS         0000000000000000  0002fb60
       0000000000001940  0000000000000000           0     0     16
  [17] .debug_info       PROGBITS         0000000000000000  000314a0
       0000000000067414  0000000000000000           0     0     1
  [18] .debug_abbrev     PROGBITS         0000000000000000  000988b4
       000000000000ce5b  0000000000000000           0     0     1
  [19] .debug_line       PROGBITS         0000000000000000  000a570f
       0000000000017c28  0000000000000000           0     0     1
  [20] .debug_frame      PROGBITS         0000000000000000  000bd338
       0000000000003370  0000000000000000           0     0     8
  [21] .debug_str        PROGBITS         0000000000000000  000c06a8
       0000000000015984  0000000000000001  MS       0     0     1
  [22] .debug_loc        PROGBITS         0000000000000000  000d602c
       0000000000023000  0000000000000000           0     0     1
  [23] .debug_ranges     PROGBITS         0000000000000000  000f902c
       000000000000a660  0000000000000000           0     0     1
  [24] .symtab           SYMTAB           0000000000000000  00103690
       0000000000008460  0000000000000018          25   697     8
  [25] .strtab           STRTAB           0000000000000000  0010baf0
       0000000000005dfa  0000000000000000           0     0     1
  [26] .shstrtab         STRTAB           0000000000000000  001118ea
       000000000000010f  0000000000000000           0     0     1
```

jhedberg commented:
@andyross after looking at this and trying to make it work with modified linker scripts, I'm starting to think it's after all better to make zefi support copying and mem-zeroing of arbitrary sized buffers, i.e. we'd essentially need to implement memset and memcpy into zefi.c and start storing byte counts for sizes and offsets instead of qwords in the bookkeeping structs.

jhedberg commented:
Actually, I'm currently testing a version which uses words instead of qwords, since that's still simpler than arbitrary-sized copies.

jhedberg commented:
@andyross now downgraded from qwords to words, however I did still need to keep the zero padding for the zeroed segments since e.g. the userspace test resulted in something that wasn't even 4-byte aligned.

andyross commented:
Seems like if we're really running into trouble here, it would make more sense just to do the copies with bytes directly instead of playing games with padding, alignment and linkage.  This really isn't a very large copy in practice (though the bss zero gets pretty hefty -- there are a LOT of page tables in that area).  If we really decide we care about performance later we can write a properly-prefix-and-suffix-aware memcpy/memset and put it into zefi.c
